V4 Program API - Placeholders and Control Flow (#1633)
rigetti/pyquil: CHANGELOG.md

- Python 3.7 is no longer supported.
- The environment variable overrides for `quilc` and `QVM` URLs have been renamed to `QCS_APPLICATIONS_QUILC_URL` and `QCS_APPLICATIONS_QVM_URL`, respectively.
- The `QuantumComputer`'s `run` method now takes an optional `memory_map` parameter. This mapping takes memory region names to a list of values to use for a run. This replaces the ability to use `write_memory` on `Program`s.

> - `PRAGMA` instructions can no longer have a directive that conflicts with a Quil keyword. If you were using directives like `Delay` or `Fence`, consider using the Quil-T instructions instead.

Reviewer: ```suggestion
- `PRAGMA` instructions can no longer have a directive that conflicts with a Quil keyword. If you were using directives like `DELAY` or `FENCE`, consider using the respective Quil-T instructions instead.
```

- `DefGate` and the other gate definition instructions will no longer accept names that conflict with Quil keywords.
- `Program#get_qubits()` will raise a `TypeError` if any of the qubits in the program are not a fixed index.

> - `QubitPlaceholders` can no longer be used in Pragma's

Reviewer: ```suggestion
- `QubitPlaceholders` can no longer be used in `PRAGMA` instructions.
```

> - A `Program`s `LabelPlaceholder`s are no longer resolved automatically when getting it's instructions. Use the `resolve_label_placeholders` method to do it explicitly. Note that the `if_then` and `while_do` methods will add `LabelPlaceholder`s to your program.

Reviewer: ```suggestion
- A `Program`s `LabelPlaceholder`s are no longer resolved automatically when getting its instructions. Use the `resolve_label_placeholders` method to do it explicitly. Note that the `if_then` and `while_do` methods will add `LabelPlaceholder`s to your program.
```

- There may be some minor differences in how instructions are converted to a Quil string. These differences should only be cosmetic and should not affect the behavior of a program. However, they may break unit tests or other code that rely on specific formatting of programs.
- The `pyquil.quil.get_default_qubit_mapping` function for getting a mapping of `QubitPlaceholders` to resolved indices has been removed. Generating a default mapping is handled automatically by the placeholder resolving methods.

> - The `JumpConditional` base class has been removed, used `JumpWhen` and/or `JumpUnless` directly instead.

Reviewer: ```suggestion
- The `JumpConditional` base class has been removed, use `JumpWhen` and/or `JumpUnless` directly instead.
```

- The `Program` class automatically sorts `DECLARE` instructions to the top of the Program when converting to Quil.

> - ``QubitPlaceholders`` are no longer supported as `Pragma` arguments.

Reviewer: duplicate

### Improvements and Changes

> - Function `pyquil.quilatom.subostitute()` now supports substitution of classical `MemoryReference`

Reviewer 2: スボー？

Author: Not sure what happened there, probably a vim motion gone wrong.

Reviewer: ```suggestion
- Function `pyquil.quilatom.substitute()` now supports substitution of classical `MemoryReference`
```
